crates/ruff_python_formatter/src/statement/suite.rs

```diff
@@ -142,18 +142,22 @@ impl FormatRule<Suite, PyFormatContext<'_>> for FormatSuite {
             )
         };
 
+        let mut preceding_comments = comments.leading_dangling_trailing(preceding);
+
         while let Some(following) = iter.next() {
+            let following_comments = comments.leading_dangling_trailing(following);
+
             // Add empty lines before and after a function or class definition. If the preceding
             // node is a function or class, and contains trailing comments, then the statement
             // itself will add the requisite empty lines when formatting its comments.
-            if (is_class_or_function_definition(preceding)
-                && !comments.has_trailing_own_line(preceding))
-                || is_class_or_function_definition(following)
+            if is_class_or_function_definition(following)
+                || (is_class_or_function_definition(preceding)
+                    && !preceding_comments.has_trailing_own_line())
```

Comment:
I preferred testing for preceding first because it better fits the preceding before following mental model
```suggestion
            if (is_class_or_function_definition(preceding)
                    && !preceding_comments.has_trailing_own_line())
                || is_class_or_function_definition(following)
```